### DISCUSSION TITLE
How to run chezmoi add in a script?

### DISCUSSION THREAD

**Discussion by agoodshort** [twpayne/chezmoi#2831] [Q&A] [answered]
I would like to run `chezmoi add` in one of my script generating a Brewfile (Homebrew "lock file"), but when doing so I face this message: `chezmoi: timeout obtaining persistent state lock, is another instance of chezmoi running?`

I feel this should be easily done, but for some reason, I can't find the way to achieve this nicely (without having my script to `cp` in `~/.local/share/chezmoi`).

```bash
#!/bin/bash
if [[ -n $HOMEBREW_BUNDLE_FILE ]]; then
    echo "Generating Brewfile..."
    brew bundle dump --force --describe
    echo "Adding Brewfile to chezmoi..."
    chezmoi add $HOMEBREW_BUNDLE_FILE
fi
```

Any suggestions?

**Comment by twpayne** [comment, marked as answer]
This violates chezmoi's core assumption that chezmoi's source state (i.e. the contents of `~/.local/share/chezmoi`) does not change while chezmoi is running.

What is the exact workflow that you want to achieve?

**Reply by agoodshort** [reply]
**Edit**: While writing the issue statement below, I realized that I might not need to rely on chezmoi at all to generate my `Brewfile` and could simply define a function in my shell to run `brew bundle ...` after `brew install ...`. I guess I was too obsessed to use `chezmoi` for something that did not require it. I will update here when I have it done and tested.

Here is my scenario:
- I have 3 different machines (2 personal and 1 for work). I use Homebrew for each of them.
- I want their `Brewfile` in [.config/homebrew](https://github.com/goodshort/dotfiles/tree/605d2a2df363e12ec55f6a415714dffc8c47ca09/home/dot_config/homebrew) so that, whichever machine I am on, I can easily refer to what is installed on the others
- I have an additional `Brewfile_Basics` which gets installed through a chezmoi script whenever changed (i.e. [run_onchange_after_brew-bundle-basics.sh.tmpl](https://github.com/goodshort/dotfiles/blob/605d2a2df363e12ec55f6a415714dffc8c47ca09/home/.chezmoiscripts/run_onchange_after_brew-bundle-basics.sh.tmpl))

My issue:
- When I install something with brew on a machine (e.g. `brew install lazygit`), it doesn't automatically update the existing `Brewfile` for this machine, so I need to run `brew bundle dump --force` after installing it and I easily forget to do so. _(but... thinking of it now, I guess I could just automate that... 🤔)_
- With the script above, I get the bundle automatically generated, but then I have to run `chezmoi re-add` to push it to chezmoi's source state, hence running `chezmoi re-add` twice...

**Reply by agoodshort** [reply]
Based on my above reply, I decided to define a function in my bash to automatically run `brew bundle dump ...` when installing something. Then I simply track the local changes with `chezmoi status` and use `chezmoi re-add` if I'm happy with the generated `Brewfile` (I don't want to `chezmoi add` right away as I might uninstall the formulae/cask a few minutes later).

Here is the simplified result for readability: 

```bash
whichbrew=$(which brew)

function brew() {
    if [[ $1 == "install" ]]; then
        $whichbrew "$@"
        $whichbrew bundle dump --force --describe # will generate Brewfile in current directory
    else
        $whichbrew "$@"
    fi
}
```

If someone is interested, [here is actual function](https://github.com/goodshort/dotfiles/blob/ad6cc1016eb7d685ad5d9226dc5398075091c52b/home/dot_config/zsh/function.zsh) I use in my `.zshrc` (there are some additional echo and env variable checks there)

**Reply by twpayne** [reply]
Great, thanks for the follow-up.
